### DISCUSSION TITLE
last()  and first() function

### DISCUSSION THREAD

**Discussion by ac4wong** [duckdb/duckdb#4871] [Q&A] [answered]
In a group by, does the order of the dataset preserves for the last() or first()?  In other words, will the two queries below return the same result set (last price of the month per ticker)

I noticed the first query is a lot faster but not confident that it's doing what I intend for. 

_select datetrunc('month', date), ticker, last(price)
from 
(
select date, ticker, price
from table
order by date
)
group by datetrunc('month', date), ticker_

vs

_select datetrunc('month', date), ticker, last (price order by date)
from table
group by datetrunc('month', date), ticker_

**Comment by Mytherin** [comment, marked as answer]
I think you are looking for `arg_max` or `arg_min`. `arg_max(price, date)` will give you the value of `price` for the row with the highest `date`, e.g.:

```sql
create table dates as select * from (values (date '1992-01-01', 'A'), (date '1992-01-02', 'B')) AS t(date, id);
select month(date) as m, arg_min(id, date), arg_max(id, date) from dates group by m;
┌───┬─────────────────────────────────┬─────────────────────────────────┐
│ m │ arg_min(id, CAST(date AS DATE)) │ arg_max(id, CAST(date AS DATE)) │
├───┼─────────────────────────────────┼─────────────────────────────────┤
│ 1 │ A                               │ B                               │
└───┴─────────────────────────────────┴─────────────────────────────────┘
```

**Reply by ac4wong** [reply]
oh this is great. Exactly what I need.

**Reply by hmms117** [reply]
For arg_max and _arg_min, what happens if two values in the second argument column are identical? Will the first value be taken? (Eg two identical dates)
Can one ensure the same row will be picked by using ORDER BY as with first() and last()?
Thanks!

**Reply by Mytherin** [reply]
Yes, you can use `ORDER BY` to resolve ambiguities, e.g.:

```sql
create table vals(i int, j int);
insert into vals values (1, 1), (1, 1000), (2, 10), (2, 100);
select arg_max(j, i order by j desc) from vals;
┌───────────────────────────────┐
│ arg_max(j, i ORDER BY j DESC) │
│             int32             │
├───────────────────────────────┤
│                           100 │
└───────────────────────────────┘
select arg_max(j, i order by j asc) from vals;
┌──────────────────────────────┐
│ arg_max(j, i ORDER BY j ASC) │
│            int32             │
├──────────────────────────────┤
│                           10 │
└──────────────────────────────┘
```

**Reply by hmms117** [reply]
Inside the function? Wow I did not know that was possible. Thanks! 🙏

**Reply by hmms117** [reply]
Hi. It turns out the ORDER BY causes my query (on a large 800m row table) to become about 100x slower. Would an ORDER BY in the SELECT statement help? Or pre-ordering the table first? 

Thanks

**Comment by l1t1** [comment]
arg_min function documents are missing in https://duckdb.org/docs/sql/aggregates
